api/kroxylicious-filter-api/src/main/templates/Kproxy/SpecificFilterArrayInvoker.ftl

```diff
@@ -0,0 +1,172 @@
+<#--
+
+    Copyright Kroxylicious Authors.
+
+    Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+
+-->
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ${outputPackage};
+
+import java.util.Arrays;
+import java.util.OptionalInt;
+
+import org.apache.kafka.common.message.ApiMessageType;
+import org.apache.kafka.common.message.RequestHeaderData;
+import org.apache.kafka.common.message.ResponseHeaderData;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ApiMessage;
+
+/**
+ * Invoker for KrpcFilters that implement any number of Specific Message interfaces (for
+ * example {@link io.kroxylicious.proxy.filter.AlterConfigsResponseFilter}.
+ */
+class SpecificFilterArrayInvoker implements FilterInvoker {
+
+    private final FilterInvoker[] requestInvokers;
+    private final FilterInvoker[] responseInvokers;
+
+    SpecificFilterArrayInvoker(KrpcFilter filter) {
+        OptionalInt maybeMaxId = Arrays.stream(ApiMessageType.values()).mapToInt(ApiMessageType::apiKey).max();
+        if (maybeMaxId.isEmpty()) {
+            throw new IllegalStateException("no maximum id found");
+        }
+        int arraySize = maybeMaxId.getAsInt() + 1;
+        this.requestInvokers = new FilterInvoker[arraySize];
+        this.responseInvokers = new FilterInvoker[arraySize];
```

Comment:
I wonder if it's worth delaying allocation until we know that the filter is actually both a request and response filter?

Reply:
I've made another implementation (to check the performance) that uses a constant array in this case to save memory. And also instead of using nulls in the array it puts a handleNothing invoker in for unhandled ids, so we can remove the null checks since every message id should have an invoker in the array. I think it's worth minimising the null cases.